rando_modules/modify_entrances.py

```diff
@@ -447,23 +435,217 @@ def get_partner_upgrade_shuffle(
+
+
+def set_starbeam_requirements(
+    world_graph: dict,
+    spirits_needed: int,
+    powerstars_needed: int,
+) -> dict:
+    """
+    Returns the modified world graph itself, modified to set the spirits
+    and power stars required to collect the Star Beam item location.
+    """
+    added_requirements = []
+
+    if spirits_needed > 0:
+        added_requirements.append([{"starspirits": spirits_needed}])
+    if powerstars_needed > 0:
+        added_requirements.append([{"powerstars": powerstars_needed}])
+
+    # find Star Beam edge and modify its requirements
+    for index, entrance in enumerate(world_graph["HOS_05/0"]["edge_list"]):
+        if (    entrance["to"]["map"] == "HOS_05"
+            and entrance["to"]["id"] == "GiftA"
+        ):
+            world_graph["HOS_05/0"]["edge_list"][index]["reqs"].extend(added_requirements)
+            #print(world_graph["HOS_05/0"]["edge_list"][index]["reqs"])
+            break
+
+    # add requirements to chapter 8 jr. troopa and both bowser battles, so the
+    # logic doesn't require going through them too early.
+    # this will add onto requirements placed by set_starway_requirements
+
+    ## ch8 jr. troopa
+    for index, entrance in enumerate(world_graph["KPA_83/0"]["edge_list"]):
+        if (    entrance["to"]["map"] == "KPA_83"
+            and entrance["to"]["id"] == 1
+        ):
+            world_graph["KPA_83/0"]["edge_list"][index]["reqs"].extend(added_requirements)
+            break
+    for index, entrance in enumerate(world_graph["KPA_83/1"]["edge_list"]):
+        if (    entrance["to"]["map"] == "KPA_83"
+            and entrance["to"]["id"] == 0
+        ):
+            world_graph["KPA_83/1"]["edge_list"][index]["reqs"].extend(added_requirements)
+            break
+    ## hallway bowser
+    for index, entrance in enumerate(world_graph["KKJ_13/0"]["edge_list"]):
+        if (    entrance["to"]["map"] == "KKJ_13"
+            and entrance["to"]["id"] == 1
+        ):
+            world_graph["KKJ_13/0"]["edge_list"][index]["reqs"].extend(added_requirements)
+            break
+    for index, entrance in enumerate(world_graph["KKJ_13/1"]["edge_list"]):
+        if (    entrance["to"]["map"] == "KKJ_13"
+            and entrance["to"]["id"] == 0
+        ):
+            world_graph["KKJ_13/1"]["edge_list"][index]["reqs"].extend(added_requirements)
+            break
+    ## final bowser
+    for index, entrance in enumerate(world_graph["KKJ_25/0"]["edge_list"]):
+        if (    entrance["to"]["map"] == "KKJ_25"
+            and entrance["to"]["id"] == 0
+        ):
+            world_graph["KKJ_25/0"]["edge_list"][index]["reqs"].extend(added_requirements)
+            break
+
+    # also add requirements to chapter 8 Koopatrol and lava switch battles, so
+    # the logic doesn't require going through them too early
+    # This is far from ideal, but an okay-ish temporary solution to having
+    # to battle chapter 8 enemies way too early
```

Comment:
Isn't this just a repeat of the code around line 520? Do we really need it?

Reply:
If we don't have the same code again here, then we would ignore the star beam requirements when setting up logic to fight the chapter 8 battles.
That would be a problem if someone set star way to 0 spirits and 0 power stars, and star beam to 7 spirits 120 stars, making it so the chapter 8 battles are in logic in sphere zero